### PR TITLE
Changes to Engine:

### PR DESCRIPTION
**Pull request by marco-scoppetta**
+ added in RESTLoader a thread scheduled at fixed rate to check if possible 
          to start post-processing
    + added getAvailableThreads in ConfigProperties
    + minor refactoring of BackgroundTasks

### CODE REVIEW

File: conf/main/mindmaps-engine.properties [hunk 1]
```diff
@@ -37,12 +37,13 @@ loader.repeat-commits=5
...
 halBuilder.degree = 1
 
 #Background tasks Config
 backgroundTasks.post-processing-delay=300000
-backgroundTasks.maintenance-iteration=1000000
-loader.threads=8
+backgroundTasks.time-lapse=60000
+
+#Loader Config
+loader.threads=0
```
Comment (fppt) [inline conf/main/mindmaps-engine.properties]: 0 threads ?

File: mindmaps-engine/src/main/java/io/mindmaps/api/ImportController.java [hunk 2]
```diff
@@ -132,6 +133,7 @@ void importDataFromFile(String dataFile) throws IOException {
         loader.waitToFinish();
         QueryParser.create().parsePatternsStream(new FileInputStream(dataFile)).forEach(pattern -> consumeRelation(pattern.admin().asVar()));
         loader.waitToFinish();
+        BackgroundTasks.getInstance().forcePostprocessing();
```
Comment (fppt) [inline mindmaps-engine/src/main/java/io/mindmaps/api/ImportController.java]: This means that after each data import from a file we run BackgroundTasks ? Could this accidentally result in 2 or more BackgroundTasks being run ?

File: mindmaps-engine/src/main/java/io/mindmaps/loader/RESTLoader.java [hunk 3]
```diff
@@ -82,26 +82,36 @@ public int getLoadingJobs() {
...
     }
 
+
     public static synchronized RESTLoader getInstance() {
         if (instance == null) instance = new RESTLoader();
         return instance;
     }
 
+    private void startPeriodPostProcessingCheck(){
+        Executors.newSingleThreadScheduledExecutor()
```
Comment (fppt) [inline mindmaps-engine/src/main/java/io/mindmaps/loader/RESTLoader.java]: What is Data Import started a BackgroundTask, would this interfere ?

File: mindmaps-engine/src/main/java/io/mindmaps/util/ConfigProperties.java [hunk 4]
```diff
@@ -52,16 +58,39 @@
     public static final String CURRENT_DIR_SYSTEM_PROPERTY = "mindmaps.dir";
     public static final String CONFIG_FILE_SYSTEM_PROPERTY = "mindmaps.conf";
 
+    private final Logger LOG = LoggerFactory.getLogger(ConfigProperties.class);
 
+    private final int MAX_NUMBER_OF_THREADS = 120;
     private Properties prop;
     private static ConfigProperties instance = null;
     private String configFilePath = null;
+    private int numOfThreads = -1;
```
Comment (fppt) [inline mindmaps-engine/src/main/java/io/mindmaps/util/ConfigProperties.java]: Shouldn't this be Atomic maybe ? Multiple threads are accessing it no ?

File: conf/main/mindmaps-engine.properties (hunk 1 above)
Comment (marco-scoppetta) [inline conf/main/mindmaps-engine.properties]: 0 threads means that Engine will use all the available cores of the current JVM

File: mindmaps-engine/src/main/java/io/mindmaps/util/ConfigProperties.java (hunk 4 above)
Comment (marco-scoppetta) [inline mindmaps-engine/src/main/java/io/mindmaps/util/ConfigProperties.java]: worst case: one thread will see it as -1 even if another one is modifying it and will force ConfigProperties to recompute the number of threads, which should always be the same. 
But I can make it atomic if it makes more sense

File: mindmaps-engine/src/main/java/io/mindmaps/postprocessing/BackgroundTasks.java [hunk 5]
```diff
@@ -78,24 +80,13 @@ public void forcePostprocessing() {
...
-            while (RESTLoader.getInstance().getLoadingJobs() != 0) {
-                try {
-                    wait();
-                } catch (InterruptedException e) {
-                    LOG.error("Error while waiting for lock", e);
-                }
-            }
-            LOG.info("QueueManager safely locked. Proceeding with maintenance");
+        if (!isRunning.get()) {
+            LOG.info("Starting maintenance.");
```
Comment (marco-scoppetta) [inline mindmaps-engine/src/main/java/io/mindmaps/postprocessing/BackgroundTasks.java]: @fppt here the check to determine wether the post processing is already running

**Pull request merged**